server_src/routes/plugins/octoprintRegistration.js

```diff
@@ -0,0 +1,165 @@
+const { Runner } = require("../../runners/state");
+const { ensureAuthenticated } = require("../../config/auth");
+const { Validator } = require("node-input-validator");
+const router = express.Router();
+
+async function handleInputValidation(req, rules, res) {
+  const v = new Validator(req.body, rules);
+  const matched = await v.check();
+  if (!matched) {
+    res.status(422).send(v.errors);
+    return false;
+  } else {
+    return v.inputs;
+  }
+}
+
+const checkInputRules = {
+  persistenceUuid: "required",
+  deviceUuid: "required",
+  host: "string",
+  port: "required|numeric",
+  docker: "required|boolean",
+  allowCrossOrigin: "required|boolean"
+};
+/**
+ * W.I.P handshake initiation endpoint
+ */
+router.get("/announce", async (req, res) => {
+  // First validate input
+  let validInput = await handleInputValidation(req, checkInputRules, res);
+  if (!validInput) return;
+  else validInput.secured = req.secure;
+
+  let deviceHost = validInput.host;
+  if (!deviceHost) {
+    deviceHost = req.socket.remoteAddress;
+  }
+  const prefix = req.secure ? "https" : "http";
+  const deviceUrl = `${prefix}://${deviceHost}:${validInput.port}`;
+
+  // Check registration - whether it doesnt exists
+  let foundRegistration =
+    await OctoprintRegistrationService.findAnnouncedDevice(
+      validInput.deviceUuid,
+      validInput.persistenceUuid
+    );
+
+  if (!foundRegistration) {
+    foundRegistration = await OctoprintRegistrationService.create({
+      isRegistered: false,
+      deviceUuid: validInput.deviceUuid,
+      persistenceUuid: validInput.persistenceUuid,
+      host: deviceHost,
+      port: validInput.port,
+      secured: req.secure,
+      docker: validInput.docker,
+      allowCrossOrigin: validInput.allowCrossOrigin
+    });
+  } else {
+    foundRegistration = await OctoprintRegistrationService.updateMetaData(
+      foundRegistration._id,
+      validInput
+    );
+  }
+
+  // Query the OctoPrint API
+  // const response = await nodeFetch(`${deviceUrl}/api/version`);
+
+  res.send({
+    alive: true,
+    dateAdded: foundRegistration.dateAdded,
+    deviceUrl: deviceUrl
+  });
+});
+
+const listAnnouncementValidationRules = {
+  onlyNotRegistered: "required|boolean"
+};
+/**
+ * List all the announced printers
+ * {isRegistered} Filter on non-registered printers or return all printers
+ */
+router.get("/list-announcements", async (req, res) => {
+  // First validate input
+  const validInput = await handleInputValidation(
+    req,
+    listAnnouncementValidationRules,
+    res
+  );
+  if (!validInput) return;
+
+  const registrations = await OctoprintRegistrationService.list(
+    validInput.onlyNotRegistered
+  );
+
+  res.send({
+    registrations
+  });
+});
+
+const registrationValidationRules = {
+  persistenceUuid: "required",
+  deviceUuid: "required",
+  deviceUrl: "required",
+  deviceApplicationKey: "required"
+};
```

Comment:
@NotExpectedYet this is what I meant with input validation.

Reply:
Aye I need to have a solid look at it, the package looks nice!